### PR TITLE
Fix/expires timeout configuration

### PR DESCRIPTION
**Pull request by wojtekzyla**
# Description

Add configuration of tasks expiration time

- [x] Refactor/improvement

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Manual and usnit testing

## Checklist

- [x] My commit message is [conventional](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] I have run pre-commit on all files before creating the PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: charts/splunk-connect-for-snmp/values.yaml [hunk 1]
```diff
@@ -175,6 +175,9 @@ scheduler:
   podAntiAffinity: soft
   nodeSelector: {}
 
+  # tasks expirations time in seconds
+  tasksExpiryTime: 180
```
Comment (omrozowicz-splunk) [inline charts/splunk-connect-for-snmp/values.yaml:179]: maybe let's not change the default one, I think the default one was 60 seconds

File: docs/configuration/scheduler-configuration.md [hunk 2]
```diff
@@ -69,3 +69,11 @@ scheduler:
 The one obligatory field for the host configuration is `address`. If `port` isn't configured its default value is `161`. 
 Other fields that can be modified here are: `community`, `secret`, `version`, `security_engine`.
 However, if they remain unspecified in the host configuration, they will be derived from the inventory record regarding this specific group.
+
+### Define tasks expiry time
+
+Define time in second after which polling or walk tasks, that haven't been picked up by the worker, will be revoked.
```
Comment (omrozowicz-splunk) [inline docs/configuration/scheduler-configuration.md:75]: maybe let's add here the page from celery documentation?

File: splunk_connect_for_snmp/common/task_generator.py [hunk 3]
```diff
@@ -24,32 +32,44 @@ def generate_task_definition(self):
...
-                    .set(priority=1),
-                    signature("splunk_connect_for_snmp.splunk.tasks.send")
-                    .set(queue="send")
-                    .set(priority=0),
+                .set(priority=4),
+                group(
+                    signature(
+                        "splunk_connect_for_snmp.inventory.tasks.inventory_setup_poller",
+                        kwargs={
+                            "chain_of_tasks_expiry_time": self.chain_of_tasks_expiry_time
```
Comment (omrozowicz-splunk) [inline splunk_connect_for_snmp/common/task_generator.py:57]: why this parameter is also in kwargs?

**Review by omrozowicz-splunk — CHANGES_REQUESTED** [review]
Tbh, I think it might have worked better if we just sense if value of expires is different than in the previous run in migration function, and if so - we delete all the tasks and set up new ones

**Comment by wojtekzyla** [comment]
> > Tbh, I think it might have worked better if we just sense if value of expires is different than in the previous run in migration function, and if so - we delete all the tasks and set up new ones
> 
> like here:
> 
> https://github.com/splunk/splunk-connect-for-snmp/blob/1a013ea0b8d9ae4483bcdb22e24442b09cc127a1/splunk_connect_for_snmp/common/schema_migration.py#L79

Sure, however we also need to delete all existing walk tasks, as we are using `expires` variable  also in those tasks.

**Comment by omrozowicz-splunk** [comment]
> 



> > > Tbh, I think it might have worked better if we just sense if value of expires is different than in the previous run in migration function, and if so - we delete all the tasks and set up new ones
> > 
> > 
> > like here:
> > https://github.com/splunk/splunk-connect-for-snmp/blob/1a013ea0b8d9ae4483bcdb22e24442b09cc127a1/splunk_connect_for_snmp/common/schema_migration.py#L79
> 
> Sure, however we also need to delete all existing walk tasks, as we are using `expires` variable also in those tasks.

Yes, I think it's better to recreate the walks instead of providing everywhere that complicated logic

File: splunk_connect_for_snmp/common/task_generator.py (hunk 3 above)
Comment (wojtekzyla) [reply splunk_connect_for_snmp/common/task_generator.py:57]: This is because polling tasks are set up inside walk tasks. We need to somehow pass this parameter to the `inventory_setup_poller ` function, so that worker running this walk task knows what should be the `expiry` value for polling tasks.

I also tested the solution when we don't pass this parameter this way, instead we take it from the environmental variable inside a worker which sets up new polling tasks. The idea was that after this env variable is changed, the worker would restart with the new env variable value and new polling tasks would be created with updated `expiry` time. The problem was, that after running `helm upgrade` with env variable changed, that old worker still worked for a short period of time, and some of the records from the inventory didn't have `expiry` value changed while others had. 
So I decided that we will be checking `CHAIN_OF_TASKS_EXPIRY_TIME` env variable only in `loader` where it is always going to be up to date and we will pass this value to walk tasks. New polling tasks are later created inside walk task by the workers.

File: splunk_connect_for_snmp/common/schema_migration.py [hunk 4]
```diff
@@ -67,10 +67,29 @@ def migrate_database(mongo_client, periodic_obj):
...
 
         save_schema_version(mongo_client, CURRENT_SCHEMA_VERSION)
 
+    previous_expiry_time = periodic_obj.get_chain_of_task_expiry()
+    logger.info(
+        f"Chain of tasks expiry time. Previous: {previous_expiry_time}. Current: {chain_of_tasks_expiry_time}"
+    )
+    expiry_time_changed = False
+    if (
+        previous_expiry_time is None
```
Comment (omrozowicz-splunk) [inline splunk_connect_for_snmp/common/schema_migration.py:85]: why it is done as when previous expiry time is None? that indicates there's no tasks yet

File: splunk_connect_for_snmp/common/task_generator.py (hunk 3 above)
Comment (omrozowicz-splunk) [reply splunk_connect_for_snmp/common/task_generator.py:57]: after the newest changes - assuming we're always deleting all the polls and walks, can we just use env variable to set it up?

**Review by omrozowicz-splunk — CHANGES_REQUESTED** [review]
Ok, I was thinking about it one more time and I'd even simplify it even more. We can either use that get expire from current tasks from redis in every load, and if it's different -> delete walks and polls and add new ones and retrigger, or (I think better idea) trigger another check similar to migrate database, keep one record in the db and check if it has changed - if yes, delete walks, polls and recreate and retrigger walks. What do you think?

File: splunk_connect_for_snmp/common/task_generator.py [hunk 5]
```diff
@@ -24,15 +34,15 @@ def generate_task_definition(self):
...
     WALK_CHAIN_OF_TASKS = {
+        "expires": CHAIN_OF_TASKS_EXPIRY_TIME,
         "link": chain(
             signature("splunk_connect_for_snmp.enrich.tasks.enrich")
             .set(queue="poll")
             .set(priority=4),
             group(
                 signature(
-                    "splunk_connect_for_snmp.inventory.tasks.inventory_setup_poller"
+                    "splunk_connect_for_snmp.inventory.tasks.inventory_setup_poller",
```
Comment (omrozowicz-splunk) [inline splunk_connect_for_snmp/common/task_generator.py:45]: was this comma provided by pre-commit?

File: test/inventory/test_loader.py [hunk 6]
```diff
@@ -98,14 +141,26 @@ def test_walk_task_for_port_161(self):
...
         self.assertEqual("sc4snmp;192.68.0.1;walk", result["name"])
         self.assertEqual("splunk_connect_for_snmp.snmp.tasks.walk", result["task"])
         self.assertEqual("192.68.0.1", result["target"])
         self.assertEqual([], result["args"])
-        self.assertEqual({"address": "192.68.0.1", "profile": None}, result["kwargs"])
+        self.assertEqual(
+            {
+                "address": "192.68.0.1",
+                "profile": None,
+                "chain_of_tasks_expiry_time": chain_of_tasks_expiry_time,
```
Comment (omrozowicz-splunk) [inline test/inventory/test_loader.py:155]: Can you add a test when `chain_of_tasks_expiry_time` is `None`?

Comment (wojtekzyla) [reply test/inventory/test_loader.py:155]: Here `chain_of_tasks_expiry_time ` will never be `None` because it is created [here](https://github.com/splunk/splunk-connect-for-snmp/blob/fix/expires-timeout-configuration/splunk_connect_for_snmp/common/task_generator.py#L81), where this variable is set from env variable which has default value of 60 ([here](https://github.com/splunk/splunk-connect-for-snmp/blob/fix/expires-timeout-configuration/splunk_connect_for_snmp/common/task_generator.py#L6C14-L6C14).

However, there is one function which returns `None` if there are no tasks in redbeat or if existing tasks don't have an `expires` variable and this is [get_chain_of_task_expiry](https://github.com/splunk/splunk-connect-for-snmp/blob/fix/expires-timeout-configuration/splunk_connect_for_snmp/customtaskmanager.py#L118-L123). This function is used [here](https://github.com/splunk/splunk-connect-for-snmp/blob/fix/expires-timeout-configuration/splunk_connect_for_snmp/customtaskmanager.py#L55) inside `did_expiry_time_change`. Unit tests of the `get_chain_of_task_expiry` can be found [here](https://github.com/splunk/splunk-connect-for-snmp/blob/fix/expires-timeout-configuration/test/other/test_custom_task_manager.py#L234-L256) and tests if the `did_expiry_time_change` where `did_expiry_time_change` will return `None` is [here](https://github.com/splunk/splunk-connect-for-snmp/blob/fix/expires-timeout-configuration/test/other/test_custom_task_manager.py#L292-L298)

**Review by omrozowicz-splunk — APPROVED** [review]
